mullvad-daemon/src/main.rs

```diff
@@ -916,3 +945,220 @@ fn running_as_admin() -> bool {
+    }
+
+    define_windows_service!(service_main, handle_service_main);
+
+    pub fn handle_service_main(arguments: Vec<OsString>) {
+        debug!("Started as a system service.");
+
+        let _ = start_service(arguments);
+    }
+
+    fn start_service(_arguments: Vec<OsString>) -> Result<()> {
+        let windows_directory = ::std::env::var_os("WINDIR").unwrap();
+        let tunnel_log_file = PathBuf::from(windows_directory)
+            .join("Temp")
+            .join("mullvad-daemon-openvpn.log");
+
+        let resource_dir = get_resource_dir();
+        let daemon = Daemon::new(Some(tunnel_log_file), resource_dir, true)
+            .chain_err(|| "Unable to initialize daemon")?;
+        let shutdown_handle = daemon.shutdown_handle();
+
+        // Register service event handler
+        let event_handler = move |control_event| -> ServiceControlHandlerResult {
+            match control_event {
+                // Notifies a service to report its current status information to the service
+                // control manager. Always return NO_ERROR even if not implemented.
+                ServiceControl::Interrogate => ServiceControlHandlerResult::NoError,
+
+                ServiceControl::Shutdown | ServiceControl::Stop => {
+                    shutdown_handle.shutdown();
+                    ServiceControlHandlerResult::NoError
+                }
+
+                _ => ServiceControlHandlerResult::NotImplemented,
+            }
+        };
+
+        let status_handle =
+            service_control_handler::register_control_handler(SERVICE_NAME, event_handler)
+                .chain_err(|| "Failed to register a service control handler")?;
+
+        // Chicken or egg problem: `ServiceState::StartPending` is never sent because
+        // `event_handler` needs `DaemonShutdownHandle`.
+        //
+        // At the same time Daemon seems to initialize quick enough so there is no need for
+        // explicit `ServiceState::StartPending` which the app enters automatically anyway.
+        //
+        // The difference in behavior may be that when Windows puts the app into
+        // `ServiceState::StartPending` state it probably uses some system default timeout to
+        // conclude that the service has hung. In case of explicit `StartPending` we can specify the
+        // approximate time until transition to `ServiceState::Running`.
+        update_service_status(&status_handle, ServiceStateUpdate::Running).unwrap();
+
+        let result = daemon.run();
+
+        // TBD: Catch Daemon shutdown and change service status to `ServiceState::StopPending`
+
+        update_service_status(
+            &status_handle,
+            ServiceStateUpdate::Stopped(ServiceExitCode::Win32(0)),
+        ).unwrap();
+
+        result
+    }
+
+    /// Checkpoint counter for ServiceState.
+    static CHECKPOINT_COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+    #[derive(Debug)]
+    enum ServiceStateUpdate {
```

Comment:
This is just a convenience `enum`. Fields like `wait_hint` and `exit_code` only make sense in certain states, such as all pending states expect `wait_hint` and `Stop` is the only one that uses `exit_code`.